Added multilingual_back_translation

Sotwi opened the pull request:
Multilingual paraphrasing with m2m100

kaustubhdhole commented:
Adding @AbinayaM02 as reviewer

AbinayaM02 commented on transformations/multilingual_back_translation/README.md:
```diff
@@ -0,0 +1,15 @@
+Author Affiliation: [SyNaLP](https://synalp.loria.fr/)([LORIA](https://www.loria.fr/en/))
+
+## What type of a transformation is this?
+This transformation is a simple paraphraser that works on 100 different languages thanks to the m2m100 many-to-many translation model. The user only needs to speciffy the source and target language. It is possible to select the same language as source and pivot, in which case the many-to-many translation model will perform a "Direct Translation" from the source language to itslef (while this is f ...
```
The supported languages can also be added here for reference, so the users can choose the listed ones only.

Sotwi replied:
Great idea. Done!

AbinayaM02 commented on transformations/multilingual_back_translation/transformation.py:
```diff
@@ -0,0 +1,63 @@
+    def generate(self, sentence: str):
+
+        # Source to Pivot
+        self.tokenizer.src_lang = self.src_lang
+        encoded_source_sentence = self.tokenizer(sentence, return_tensors="pt")
+        generated_pivot_tokens = self.model.generate(**encoded_source_sentence, forced_bos_token_id=self.tokenizer.get_lang_id(self.pivot_lang))
+        pivot_sentence = self.tokenizer.batch_decode(generated_pivot_tokens, skip_special_tokens=True)
+    
+        #Pivot to Source
+        if self.pivot_lang != self.src_lang:
+            self.tokenizer.src_lang = self.pivot_lang
+            encoded_pivot_sentence = self.tokenizer(pivot_sentence, return_tensors="pt")
+            generated_source_tokens = self.model.generate(**encoded_pivot_sentence, forced_bos_token_id=self.tokenizer.get_lang_id(self.src_lang))
+            source_sentence = self.tokenizer.batch_decode(generated_source_tokens, skip_special_tokens=True)
+        else:
+            source_sentence = pivot_sentence
```
Just to make the code cleaner and reusable, I would suggest having a single method like below doing the translation,
```python
def source_to_target(src_lang, pivot_lang, sentence):
     # logic to translate 
```
And this method can be called in generate to translate from source to pivot language and vice versa just by switch the src_lang and pivot_lang params.

Sotwi replied:
Done!

AbinayaM02 reviewed (commented):
Have added my comments in the respective files. Please have a look.

AbinayaM02 reviewed (changes requested):
Since the transformation is multi-lingual, it would be nice if you evaluate it against some standard dataset for any specific task. Please, refer the link : https://github.com/GEM-benchmark/NL-Augmenter/tree/main/evaluation, for more information on how to evaluate your transformation.

Sotwi commented on transformations/multilingual_back_translation/README.md:
```diff
@@ -0,0 +1,18 @@
+This transformations works to increase the data for any task that has input texts.
+
+## What are the limitations of this transformation?
+The outputs dependend on the accuracy of the translation model for thegiven pair of languages.
```
Made the suggested corrections.

Sotwi replied:
I added the results of the evaluations that I managed to get. Could not apply the transformation to `roberta-base-imdb` on the IMDB Dataset. It was too time consuming.

AbinayaM02 reviewed (approved):
Hi @Sotwi : Thank you for addressing the requested changes. Here are my comments,
1. The transformation looks fine and the right interface has been chosen.
2. The transformation is generic in the sense it supports 100 languages and it's not already implemented in NL-Augmenter.
3. No new library is added but a new model is used.
4. Test cases have been added.
5. Robustness evaluation is performed for text classification.

kaustubhdhole commented on evaluation/README.md:
```diff
@@ -1,84 +1,85 @@
+| RedundantContextForQa |                           5.6 |                                                    -1.9 |
+
+### Question Generation
+### AMR-to-Text
```
Okay.. one last thing - seems this is a formatting change - might be easier to track later if you revert this file's change. Rest is good to go :)

kaustubhdhole replied:
I mean you could just add a new line

Sotwi replied:
I'm not sure what caused the whole file to change My suspects are either my file editor or pre-commit since I had just adjusted the table.
Either way, I think it's working now.
Glad to help and be part of this project.

kaustubhdhole replied:
Thanks for adding this.

kaustubhdhole commented on transformations/multilingual_back_translation/README.md:
```diff
@@ -0,0 +1,29 @@
+
+
+Angela Fan, Shruti Bhosale, Holger Schwenk, Zhiyi Ma, Ahmed El-Kishky, Siddharth Goyal, Mandeep Baines, Onur Celebi, Guillaume Wenzek, Vishrav Chaudhary, Naman Goyal, Tom Birch, Vitaliy Liptchinsky, Sergey Edunov, Edouard Grave, Michael Auli and Armand Joulin. 2021. [Beyond English-Centric Multilingual Machine Translation](https://www.jmlr.org/papers/volume22/20-1307/20-1307.pdf).
+
```
Okay, for the benefit of people reading, you might want to specify [back_translation](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/transformations/back_translation) as well as [lost_in_translation](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/transformations/lost_in_translation) in the related work and how this PR is different.

Sotwi replied:
Done, hopefully it is clear enough.

kaustubhdhole commented:
Thanks @Sotwi for addressing the concerns. This PR uses another model for back-translation and extends it to multiple languages. Backtranslation is a popular procedure and I feel its multilingual version would benefit the project. The code is excellently written and provides access to more languages. It is interesting to see large drops in performance testing ~8% and ~11%. I would suggest adding the test of `roberta-base-imdb` too. But I am happy to approve this in the current state.

Sotwi commented:
> Thanks @Sotwi for addressing the concerns. This PR uses another model for back-translation and extends it to multiple languages. Backtranslation is a popular procedure and I feel its multilingual version would benefit the project. The code is excellently written and provides access to more languages. It is interesting to see large drops in performance testing ~8% and ~11%. I would suggest adding the test of `roberta-base-imdb` too. But I am happy to approve this in the current state.

Hello @kaustubhdhole, I had not computed `roberta-base-imdb` because it was taking too long to compute. I later realized that the transformation was not making use of the GPU. I have added a couple lines of code to make sure the transformations uses the GPU when it is available and then computed the results for `roberta-base-imdb` (92.0 ie. -3.0). How do I apply those changes now that the branch is merged?
